### PR TITLE
chore(tracer/writer): improve writing exception

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
The current exception prints out a loud stack trace that contains a lot of unhelpful information for both the developers and users of the library.

```
ERROR:ddtrace.internal.writer:failed to send traces to Datadog Agent at http://localhost:8126/v0.5/traces
Traceback (most recent call last):
  File "/Users/kverhoog/dev/dd-trace-py/.venv/lib/python3.10/site-packages/tenacity/__init__.py", line 409, in __call__
    result = fn(*args, **kwargs)
  File "/Users/kverhoog/dev/dd-trace-py/ddtrace/internal/writer.py", line 447, in _send_payload
    response = self._put(payload, headers)
  File "/Users/kverhoog/dev/dd-trace-py/ddtrace/internal/writer.py", line 399, in _put
    self._conn.request("PUT", self._endpoint, data, headers)
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/http/client.py", line 1282, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/http/client.py", line 1328, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/http/client.py", line 1277, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/http/client.py", line 1037, in _send_output
    self.send(msg)
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/http/client.py", line 975, in send
    self.connect()
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/http/client.py", line 941, in connect
    self.sock = self._create_connection(
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/socket.py", line 845, in create_connection
    raise err
  File "/Users/kverhoog/.pyenv/versions/3.10.3/lib/python3.10/socket.py", line 833, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 61] Connection refused

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/kverhoog/dev/dd-trace-py/ddtrace/internal/writer.py", line 564, in flush_queue
    self._retry_upload(self._send_payload, encoded, n_traces)
  File "/Users/kverhoog/dev/dd-trace-py/.venv/lib/python3.10/site-packages/tenacity/__init__.py", line 406, in __call__
    do = self.iter(retry_state=retry_state)
  File "/Users/kverhoog/dev/dd-trace-py/.venv/lib/python3.10/site-packages/tenacity/__init__.py", line 363, in iter
    raise retry_exc from fut.exception()
tenacity.RetryError: RetryError[<Future at 0x107799210 state=finished raised ConnectionRefusedError>]
```

Instead, the useful information can be included in the exception and the stack trace can be omitted.

```
ERROR:ddtrace.internal.writer:failed to send 1 traces to Datadog Agent at http://localhost:8126/v0.5/traces after 3 retries ([Errno 61] Connection refused)
```

## Reviewer Checklist
- [x] Title is accurate.
- [x] Description motivates each change.
- [x] No unnecessary changes were introduced in this PR.
- [x] Avoid breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Tests provided or description of manual testing performed is included in the code or PR.
- [x] Release note has been added and follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines), or else `changelog/no-changelog` label added.
- [x] All relevant GitHub issues are correctly linked.
- [x] Backports are identified and tagged with Mergifyio.

### CODE REVIEW

File: ddtrace/internal/writer.py [hunk 1]
```diff
@@ -569,7 +569,13 @@ def flush_queue(self, raise_exc=False):
                 if raise_exc:
                     e.reraise()
                 else:
-                    log.error("failed to send traces to Datadog Agent at %s", self._agent_endpoint, exc_info=True)
+                    log.error(
+                        "failed to send %d traces to Datadog Agent at %s after %d retries (%s)",
```
Comment (brettlangdon) [inline ddtrace/internal/writer.py:573]: should we be more explicit that the traces are dropped?

Comment (Kyle-Verhoog) [reply ddtrace/internal/writer.py:573]:
```suggestion
                        "failed to send, dropping %d traces to Datadog Agent at %s after %d retries (%s)",
```
?

Comment (brettlangdon) [reply ddtrace/internal/writer.py:573]: yeah, I like it

**Review by brettlangdon — APPROVED** [review]

**Review by brettlangdon — DISMISSED** [review]

**Comment by majorgreys** [comment]
Is there a scenario where we would want to enable printing a stack trace?

**Comment by brettlangdon** [comment]
> Is there a scenario where we would want to enable printing a stack trace?

I don't think so. We know the block of code where it is coming from, it isn't really useful for customers.

Unless we want to add a debug log for the exception?

**Review by brettlangdon — DISMISSED** [review]

File: releasenotes/notes/writer-exception-7181ac68fbb1a3d0.yaml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+---
+other:
+  - |
+    tracer: The exception logged for writing errors no longer includes a long,
```
Comment (mabdinur) [inline releasenotes/notes/writer-exception-7181ac68fbb1a3d0.yaml:4]: ?
```suggestion
    tracing: The exception logged for writing errors no longer includes a long,
```

File: tests/utils.py [hunk 3]
```diff
@@ -1014,6 +1014,11 @@ def __eq__(self, other):
         return isinstance(other, int)
 
 
+class AnyExc(object):
```
Comment (mabdinur) [inline tests/utils.py:1017]: I like this 👍

File: ddtrace/internal/writer.py [hunk 4]
```diff
@@ -569,7 +569,13 @@ def flush_queue(self, raise_exc=False):
                 if raise_exc:
                     e.reraise()
                 else:
-                    log.error("failed to send traces to Datadog Agent at %s", self._agent_endpoint, exc_info=True)
+                    log.error(
+                        "failed to send, dropping %d traces to Datadog Agent at %s after %d retries (%s)",
+                        n_traces,
+                        self._agent_endpoint,
+                        e.last_attempt.attempt_number,
+                        e.last_attempt.exception(),
```
Comment (mabdinur) [inline ddtrace/internal/writer.py:577]: Question: Would this log a tenacity retry error or the original error?

**Review by mabdinur — COMMENTED** [review]
Left some non blocking comments. Feel free to resolve and I'll 👍

File: releasenotes/notes/writer-exception-7181ac68fbb1a3d0.yaml (hunk 2 above)
Comment (Kyle-Verhoog) [reply releasenotes/notes/writer-exception-7181ac68fbb1a3d0.yaml:4]: ah yes! thank you

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**